File: frontend/src/lib/sharePoint.ts
/**
 * Integración con SharePoint / Microsoft Graph (importación puntual).
 * MSAL (popup, scope Sites.Read.All) → access_token → exploramos
 * sitio → biblioteca de documentos (drive) → carpetas con Graph, y el backend
 * descarga los archivos elegidos.
 *
 * Comparte el Client ID con OneDrive (misma App registration de Azure); lo que
 * cambia es el permiso de Graph: SharePoint necesita Sites.Read.All (delegado),
 * que normalmente requiere consentimiento del admin del tenant.
 */
import { PublicClientApplication } from "@azure/msal-browser";
import {
  getOneDriveClientId,
  setOneDriveClientId,
  type DriveItem,
} from "./oneDrive";

const GRAPH = "https://graph.microsoft.com/v1.0";
const SCOPES = ["Sites.Read.All"];

export type { DriveItem };

/** Un sitio de SharePoint. */
export interface SiteItem {
  id: string;
  name: string;
  webUrl?: string;
}

/** Una biblioteca de documentos (drive) dentro de un sitio. */
export interface DriveLib {
  id: string;
  name: string;
}

// El Client ID es el mismo que OneDrive (misma App registration de Azure).
export function getSharePointClientId(): string | undefined {
  return getOneDriveClientId();
}

export function setSharePointClientId(clientId: string): void {
  setOneDriveClientId(clientId);
}

export function isSharePointConfigured(): boolean {
  return Boolean(getOneDriveClientId());
}

let _pca: PublicClientApplication | null = null;
let _initClientId: string | null = null;

async function getPca(): Promise<PublicClientApplication> {
  const clientId = getSharePointClientId();
  if (!clientId) throw new Error("SharePoint no está configurado.");
  if (!_pca || _initClientId !== clientId) {
    _pca = new PublicClientApplication({
      auth: {
        clientId,
        authority: "https://login.microsoftonline.com/common",
        redirectUri: window.location.origin,
      },
      cache: { cacheLocation: "localStorage" },
    });
    await _pca.initialize();
    _initClientId = clientId;
  }
  return _pca;
}

/** Intenta un access_token SIN popup (devuelve null si requiere interacción). */
export async function getSharePointTokenSilent(): Promise<string | null> {
  const pca = await getPca();
  const accounts = pca.getAllAccounts();
  if (accounts.length === 0) return null;
  try {
    const res = await pca.acquireTokenSilent({
      scopes: SCOPES,
      account: accounts[0],
    });
    return res.accessToken;
  } catch {
    return null;
  }
}

/** Obtiene un access_token con popup. DEBE llamarse desde un gesto del usuario
 * (clic): los navegadores bloquean popups que no nacen de una interacción. */
export async function getSharePointToken(): Promise<string> {
  const silent = await getSharePointTokenSilent();
  if (silent) return silent;
  const pca = await getPca();
  const res = await pca.acquireTokenPopup({ scopes: SCOPES });
  return res.accessToken;
}

/** Busca sitios de SharePoint por texto (vacío = sitios más usados). */
export async function searchSites(
  token: string,
  query: string
): Promise<SiteItem[]> {
  const q = query.trim() || "*";
  const url = `${GRAPH}/sites?search=${encodeURIComponent(q)}&$top=50`;
  const resp = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!resp.ok) throw new Error(`Graph ${resp.status}`);
  const data = await resp.json();
  return (data.value || []).map(
    (s: { id: string; displayName?: string; name?: string; webUrl?: string }) => ({
      id: s.id,
      name: s.displayName || s.name || s.webUrl || s.id,
      webUrl: s.webUrl,
    })
  );
}

/** Lista las bibliotecas de documentos (drives) de un sitio. */
export async function listDrives(
  token: string,
  siteId: string
): Promise<DriveLib[]> {
  const url = `${GRAPH}/sites/${siteId}/drives?$select=id,name&$top=100`;
  const resp = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!resp.ok) throw new Error(`Graph ${resp.status}`);
  const data = await resp.json();
  return (data.value || []).map((d: { id: string; name: string }) => ({
    id: d.id,
    name: d.name,
  }));
}

/** Lista el contenido de una carpeta de una biblioteca (root si no hay folderId). */
export async function listChildren(
  token: string,
  siteId: string,
  driveId: string,
  folderId?: string
): Promise<DriveItem[]> {
  const base = `${GRAPH}/sites/${siteId}/drives/${driveId}`;
  const path = folderId
    ? `${base}/items/${folderId}/children`
    : `${base}/root/children`;
  const url = `${path}?$select=id,name,folder,file,size&$top=200`;
  const resp = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!resp.ok) throw new Error(`Graph ${resp.status}`);
  const data = await resp.json();
  return (data.value || []).map(
    (it: { id: string; name: string; folder?: unknown; size?: number }) => ({
      id: it.id,
      name: it.name,
      isFolder: Boolean(it.folder),
      size: it.size,
    })
  );
}

File: backend/app/connectors/sharepoint.py
"""Conector de SharePoint / Microsoft Graph (importación puntual).

Análogo a OneDrive (mismo Graph + MSAL), pero la jerarquía es
sitio → biblioteca de documentos (drive) → carpetas, así que cada archivo
elegido viaja con su ``site_id`` + ``drive_id``. Los archivos son binarios
reales (pdf/docx/...) → descarga directa por ``/content`` (302 a URL
prefirmada, igual que OneDrive).

Permiso de Graph necesario: ``Sites.Read.All`` (delegado) — normalmente
requiere consentimiento del admin del tenant de Azure AD.
"""
from __future__ import annotations

from pathlib import Path

import httpx

from app.ingestion.parsers import SUPPORTED_EXTENSIONS

SOURCE_SHAREPOINT = "sharepoint"

_GRAPH = "https://graph.microsoft.com/v1.0"


async def download_files(
    access_token: str, files: list[dict]
) -> tuple[list[tuple[str, bytes, str | None]], list[str], list[str]]:
    """Descarga los archivos elegidos por (site_id, drive_id, item_id) de Graph.

    Devuelve (items, rechazados, fallidos)."""
    items: list[tuple[str, bytes, str | None]] = []
    rejected: list[str] = []
    failed: list[str] = []
    headers = {"Authorization": f"Bearer {access_token}"}

    # follow_redirects: el endpoint /content responde 302 a una URL prefirmada.
    async with httpx.AsyncClient(timeout=60, follow_redirects=True) as client:
        for f in files:
            fid = f.get("id")
            site_id = f.get("site_id")
            drive_id = f.get("drive_id")
            name = f.get("name") or (fid or "documento")
            if not fid or not site_id or not drive_id:
                failed.append(name)
                continue
            if Path(name).suffix.lower() not in SUPPORTED_EXTENSIONS:
                rejected.append(name)
                continue
            try:
                resp = await client.get(
                    f"{_GRAPH}/sites/{site_id}/drives/{drive_id}/items/{fid}/content",
                    headers=headers,
                )
                resp.raise_for_status()
                ctype = resp.headers.get("content-type")
                items.append((name, resp.content, ctype))
            except Exception:  # noqa: BLE001
                failed.append(name)

    return items, rejected, failed
